lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -1295,25 +1341,88 @@ class MongoClientInterface {
      */
     deleteObjectNoVer(c, bucketName, objName, params, log, cb) {
         const masterKey = formatMasterKey(objName, params.vFormat);
-        c.findOneAndDelete({
-            _id: masterKey,
-        }, {}, (err, result) => {
+        this.internalDeleteObject(c, bucketName, masterKey, {}, log, (err, res) => {
             if (err) {
+                // Should not return an error when no object is found
+                if (err.is.NoSuchKey) {
+                    return cb(null);
+                }
                 log.error(
                     'deleteObjectNoVer: error deleting object with no version',
-                    { error: err.message });
+                    { error: err.message, bucket: bucketName, key: objName });
                 return cb(errors.InternalError);
             }
-            if (result.ok !== 1) {
+            if (res.ok !== 1) {
                 log.error(
-                    'deleteObjectNoVer: failed deleting object with no version',
-                    { error: err.message });
+                    'deleteObjectNoVer: failed to delete object',
+                    { bucket: bucketName, object: objName });
                 return cb(errors.InternalError);
             }
             return cb(null);
         });
     }
 
+    /**
+     * Flags the object before deleting it, this is done
+     * to keep object metadata in the oplog, as oplog delete
+     * events don't contain any object metadata
+     * @param {Object} collection MongoDB collection
+     * @param {string} bucketName bucket name
+     * @param {string} key Key of the object to delete
+     * @param {object} filter additional query filters
+     * @param {Logger}log logger instance
+     * @param {Function} cb callback containing error
+     * and BulkWriteResult
+     * @return {undefined}
+     */
+    internalDeleteObject(collection, bucketName, key, filter, log, cb) {
+        // filter used when finding and updating object
+        const findUpdateFilter = Object.assign({
+            _id: key,
+        }, filter);
+        // filter used when deleting object
+        const deleteFilter = Object.assign({
+            '_id': key,
+            'value.deleted': true,
+        }, filter);
+        async.waterfall([
```

Comment:
Just to make sure I understand correctly, we're only having a race condition here because:
- we first do a get+unserialize+modify using application-level setters
- we then do a bulk sequence of write+delete
- these 2 operations, done non-atomically, can be interrupted in the middle or interleaved with a change from another request

Can't the waterfall be replaced by a single `bulk-write(ordered: true, upsert: false)` with direct `$set`s of `originOp` and `deleted` (doesn't really matter that encapsulation's broken, since we're already direct-querying on `deleted`) and conditional flags - as long as we have `value.deleted: false` for the conditional update. Upon checking the bulk-write result, if the first step failed, just return an error to the caller; if the second step failed, return success (some mopping mechanism may have to run outside though as although rare, it's quite impossible to guarantee that this won't happen without transactions, as mongodb's quorum could be broken at any time).

Once we move to mongodb 4.2+, transactions can be used to do all of this atomically on sharded clusters.

Am I missing something?

(cc @francoisferrand)

Reply:
Oplog update events only show the updated fields, so if we directly update only `originOp` and `deleted`, that's all we are going to see in the oplog, and the event will look like this :
```
  "ts" : Timestamp(1666765379, 3),
    "t" : NumberLong(14),
    "h" : NumberLong(-8924670320665905898),
    "v" : 2,
    "op" : "u",
    "ns" : "metadata.arsenal-bucket",
    "ui" : UUID("09301a87-0bb6-4eb3-9ca4-11c890492d1f"),
    "o2" : {
        "_id" : "Vobject-1\u000098333234628661999999RG001  54937.0"
    },
    "wall" : ISODate("2022-10-26T06:22:59.019Z"),
    "lsid" : {
        "id" : UUID("901cb895-39af-42ce-8082-6f78f68bca8f"),
        "uid" : { "$binary" : "47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=", "$type" : "00" }
    },
    "txnNumber" : NumberLong(5),
    "stmtId" : 0,
    "prevOpTime" : {
        "ts" : Timestamp(0, 0),
        "t" : NumberLong(-1)
    },
    "postImageOpTime" : {
        "ts" : Timestamp(1666765379, 2),
        "t" : NumberLong(14)
    },
    "o" : {
        "$v" : 1,
        "$set" : {
            "value.originOp" : "s3:ObjectRemoved:Delete"
            "value.deleted" : true
        }
    }
```
This is why we first `Find` and then `Update` the whole metadata

Reply:
The reason this was implemented in 7.x is as Ronnie said [here](https://scality.slack.com/archives/CFD81UFEK/p1657214530430809?thread_ts=1657208435.407199&cid=CFD81UFEK), cause Armory is meant to be used in both 7.x and 8.x branch, so we want to keep the lib/storage folder identical in both branches to avoid issues until the code is moved to Armory